### PR TITLE
Resource Dependency Management

### PR DESCRIPTION
**Pull request by hongchaodeng**
This is an implementation based on [the design](https://github.com/crossplane/oam-kubernetes-runtime/blob/master/design/resource-dependency.md).

Covering:
- [X] DAG Manager
- [x] Components as source and sink
- [x] Traits as source and sink
- [X] Test coverage

### CODE REVIEW

**Comment by hongchaodeng** [comment]
PR is ready.
cc @resouer @wonderflow @ryanzhang-oss please review it

File: apis/core/v1alpha2/core_types.go [hunk 1]
```diff
@@ -324,3 +336,26 @@ type ApplicationConfigurationList struct {
...
+	FieldPath string `json:"fieldPath,omitempty"`
+}
+
+// DataInput specifies a data input sink to an object.
+type DataInput struct {
+	// ValueFrom specifies the value source.
+	ValueFrom DataInputValueFrom `json:"valueFrom,omitempty"`
+
+	// ToFieldPaths specifies the field paths of an object to fill passed value.
+	ToFieldPaths []string `json:"toFieldPaths,omitempty"`
```
Comment (zzxwill) [inline apis/core/v1alpha2/core_types.go:355]: How about renaming `ToFieldPaths` to `FieldPath` which is same to that of `DataOutput`? 1) This seems consistent. 2) `To` in `ToFieldPaths` sounds like `Out` in `DataOutput` which might lead to confusion as I did right now.

File: pkg/dependency/dag.go [hunk 2]
```diff
@@ -0,0 +1,55 @@
+package dependency
+
+import (
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+)
+
+// DAG is the dependency graph for an AppConfig.
+type DAG struct {
+	Sources map[string]*Source
```
Comment (ryanzhang-oss) [inline pkg/dependency/dag.go:10]: Source is an ObjectRef, why makes it a struct?

File: pkg/dependency/dagmanager.go [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+		}
+
+		if len(dag.Sources) == 0 {
+			dm.log.Debug("all dependencies satisfied", "app", app)
+			delete(dm.app2dag, app)
+		}
+	}
+}
+
+func (dm *dagManagerImpl) checkSourceReady(ctx context.Context, s *Source) (string, error) {
```
Comment (ryanzhang-oss) [inline pkg/dependency/dagmanager.go:95]: it seems that this logic belongs to the source itself.

File: pkg/dependency/dagmanager.go [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+
+			for sinkName, sink := range dag.Sinks[sourceName] {
+				dm.log.Debug("triggering sinks", "app", app, "source", sourceName, "sink", sinkName)
+				err := dm.trigger(ctx, sink, val)
+				if err != nil {
+					dm.log.Info("triggering sink failed", "errmsg", err)
+				}
+			}
+
+			delete(dag.Sources, sourceName)
```
Comment (ryanzhang-oss) [inline pkg/dependency/dagmanager.go:85]: How can we delete source here when 
1. it's still in the loop
2. sinks are not guaranteed to be ready yet

File: pkg/controller/oam/applicationconfiguration/render.go [hunk 5]
```diff
@@ -72,50 +74,97 @@ type components struct {
...
+		// Only create trait if it doesn't depend on anything.
+		if len(ct.DataInputs) == 0 {
+			traits = append(traits, *t)
+		}
+	}
 
-			t.SetOwnerReferences([]metav1.OwnerReference{*ref})
-			t.SetNamespace(ac.GetNamespace())
+	// Only create workload if it doesn't depend on anything.
+	if len(acc.DataInputs) == 0 {
```
Comment (ryanzhang-oss) [inline pkg/controller/oam/applicationconfiguration/render.go:145]: when is the workload created then?

File: pkg/controller/oam/applicationconfiguration/render.go [hunk 6]
```diff
@@ -72,50 +74,97 @@ type components struct {
...
-			}
+		addDataOutputsToDAG(dag, ct.DataOutputs, t)
+		addDataInputsToDAG(dag, ct.DataInputs, t)
 
-			// Set metadata name for `Trait` if the metadata name is NOT set.
-			if t.GetName() == "" {
-				t.SetName(acc.ComponentName)
-			}
+		// Only create trait if it doesn't depend on anything.
+		if len(ct.DataInputs) == 0 {
```
Comment (ryanzhang-oss) [inline pkg/controller/oam/applicationconfiguration/render.go:139]: when is the trait created then?

**Comment by zzxwill** [comment]
LGTM

**Comment by SevenYoung** [comment]
1. We need a place to show DAG status which support DAG debugging and progress track.
2. DAG manager should be strictly declarative, DAG can retry from failed DAG node.

**Comment by zzxwill** [comment]
No `status.connSecret` is generated on the workload `RDSInstance` of depended db component with the repo of https://github.com/hongchaodeng/oam-kubernetes-runtime on branch `dep`. Did I miss any thing to have it run successfully?

For detailed steps and demo manifests, please refer to https://github.com/oam-dev/samples/pull/43

image

**Comment by zzxwill** [comment]
I also noticed the workload of webapp was generated at the time when workload RDSInstance is generated.
```
k get configuration
NAME     LATESTCREATED   LATESTREADY   READY     REASON
webapp   webapp-km4tz                  Unknown
```

**Comment by zzxwill** [comment]
> No `status.connSecret` is generated on the workload `RDSInstance` of depended db component with the repo of https://github.com/hongchaodeng/oam-kubernetes-runtime on branch `dep`. Did I miss any thing to have it run successfully?
> 
> For detailed steps and demo manifests, please refer to [oam-dev/samples#43](https://github.com/oam-dev/samples/pull/43)
> 
> image
> 
> ```
> k get rdsinstance rdspostgresql2 -o yaml
> apiVersion: database.alibaba.crossplane.io/v1alpha1
> kind: RDSInstance
> metadata:
>   annotations:
>     crossplane.io/external-name: rdspostgresql2
>   creationTimestamp: "2020-06-24T07:35:05Z"
>   finalizers:
>   - finalizer.managedresource.crossplane.io
>   generation: 1
>   name: rdspostgresql2
>   ownerReferences:
>   - apiVersion: core.oam.dev/v1alpha2
>     blockOwnerDeletion: true
>     controller: true
>     kind: ApplicationConfiguration
>     name: knative-postgresql-appconfig
>     uid: 32c9e6be-192a-487e-be97-22088539de3b
>   resourceVersion: "879351"
>   selfLink: /apis/database.alibaba.crossplane.io/v1alpha1/rdsinstances/rdspostgresql2
>   uid: c6a689c3-fa16-4069-bcb4-c5e69c1d53bf
> spec:
>   forProvider:
>     dbInstanceClass: rds.pg.s1.small
>     dbInstanceStorageInGB: 20
>     engine: PostgreSQL
>     engineVersion: "9.4"
>     masterUsername: test123
>     securityIPList: 0.0.0.0/0
>   providerRef:
>     name: alibaba-provider
>   reclaimPolicy: Delete
>   writeConnectionSecretToRef:
>     name: alibaba-rdspostgresql-conn
>     namespace: default
> status:
>   atProvider:
>     accountReady: true
>     dbInstanceID: rm-2ze8yp1t6m52ubuam
>     dbInstanceStatus: Running
>   bindingPhase: Unbound
>   conditions:
>   - lastTransitionTime: "2020-06-24T07:38:50Z"
>     reason: Resource is available for use
>     status: "True"
>     type: Ready
>   - lastTransitionTime: "2020-06-24T07:35:15Z"
>     reason: Successfully reconciled resource
>     status: "True"
>     type: Synced
> ```

After the offline discussion with @hongchaodeng @resouer @wonderflow , the controller will provider a way to the the value of `fieldPath`. As for this issue, it will check whether the value of `status.atProvider.dbInstanceStatus` is equal to `Running`. It **MIGHT** looks like as below and `value` **MIGHT** be optional.
```
    dataOutputs:
    - name: mysql-conn-secret
      fieldPath: status.atProvider.dbInstanceStatus
      value: Running
```

**Comment by hongchaodeng** [comment]
I experience high difficulty when resolving conflicts so I rewrite the whole PR based on the master.

Moving it to https://github.com/crossplane/oam-kubernetes-runtime/pull/53

**Pull request closed without merging**